src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java

```diff
@@ -0,0 +1,165 @@
+    CompletableFuture<Void> future = new VertxCompletableFuture<>(ctx);
+    futures.add(future);
+
+    String poLineIdUpdate = messageFromEventBus.getString("poLineIdUpdate");
+    String query = String.format(PIECES_ENDPOINT, poLineIdUpdate, LIMIT);
+
+    // 1. Get all pieces for poLineId
+    getPieces(query, httpClient, okapiHeaders, logger).thenAccept(piecesCollection -> {
+      List<org.folio.rest.acq.model.Piece> listOfPieces = piecesCollection.getPieces();
+
+      // 2. Get PoLine for the poLineId which will be used to calculate PoLineReceiptStatus
+      getPoLineById(poLineIdUpdate, lang, httpClient, ctx, okapiHeaders, logger).thenAccept(poLineJson -> {
+        PoLine poLine = poLineJson.mapTo(PoLine.class);
+
+        calculatePoLineReceiptStatus(poLine, listOfPieces)
+          .thenCompose(status -> updatePoLineReceiptStatus(poLine, status, httpClient, ctx, okapiHeaders, logger))
+          .thenAccept(updatedPoLineId -> {
+            if (updatedPoLineId != null) {
+              // send event to update order status
+              updateOrderStatus(poLine, okapiHeaders);
+            }
+          })
+          .thenAccept(future::complete);
+      })
+        .exceptionally(e -> {
+          logger.error("The error getting poLine by id {}", poLineIdUpdate, e);
+          future.completeExceptionally(e);
+          return null;
+        });
+    })
+      .exceptionally(e -> {
+        logger.error("The error happened getting all pieces by poLine {}", poLineIdUpdate, e);
+        future.completeExceptionally(e);
+        return null;
+      });
+
+    // Now wait for all operations to be completed and send reply
+    completeAllFutures(ctx, httpClient, futures, message);
+  }
+
+  private void updateOrderStatus(PoLine poLine, Map<String, String> okapiHeaders) {
+    List<String> poIds = new ArrayList<>();
+    poIds.add(poLine.getPurchaseOrderId());
+    JsonObject messageContent = new JsonObject();
+    messageContent.put(OKAPI_HEADERS, okapiHeaders);
+    messageContent.put(ORDER_IDS, new JsonArray(poIds));
+    sendEvent(MessageAddress.ORDER_STATUS, messageContent);
+  }
+
+  private CompletableFuture<PoLine.ReceiptStatus> calculatePoLineReceiptStatus(PoLine poLine,
+      List<org.folio.rest.acq.model.Piece> pieces) {
+
+    if (pieces.isEmpty()) {
+      return completedFuture(poLine.getReceiptStatus());
+    } else {
+      return getPiecesQuantityByPoLineAndStatus(ReceivingStatus.EXPECTED, pieces)
+        .thenCompose(expectedQty -> calculatePoLineReceiptStatus(expectedQty, pieces))
+        .exceptionally(e -> {
+          logger.error("The expected receipt status for PO Line '{}' cannot be calculated", e, poLine.getId());
+          return null;
+        });
+    }
+  }
+
+  private CompletableFuture<ReceiptStatus> calculatePoLineReceiptStatus(int expectedPiecesQuantity,
+      List<org.folio.rest.acq.model.Piece> pieces) {
+    
+    if (expectedPiecesQuantity == 0) {
```

Comment:
@piotr-kalashuk - I think the code will never reach here because if `pieces` are empty then future will complete and returned on line 111(`pieces.isEmpty()`) above. Is there any scenario when `expectedPiecesQuantity` be 0 and pieces are not empty?